File: 155-min-stack/155-min-stack.java
class ListNode{
    int val;
    int min;
    ListNode next;
    public ListNode(int val, int min, ListNode next){
        this.val = val;
        this.min = min;
        this.next = next;
    }
}
class MinStack {
    ListNode head;
    public MinStack() {
        head = new ListNode(Integer.MAX_VALUE, Integer.MAX_VALUE, null);
    }
    
    public void push(int val) {
        head = new ListNode(val, Math.min(head.min, val), head);
    }
    
    public void pop() {
        head = head.next;
    }
    
    public int top() {
        return head.val;
    }
    
    public int getMin() {
        return head.min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155-min-stack/155-min-stack.py
class MinStack:

    def __init__(self):
        self.st1 = []
        self.st2 = []
        

    def push(self, val: int) -> None:
        if len(self.st2) ==0 or val <= self.st2[-1]:
            self.st2.append(val)
        self.st1.append(val)
        
    def pop(self) -> None:
        if self.st1[-1] == self.st2[-1]:
            self.st2.pop()
        self.st1.pop()

    def top(self) -> int:
        return self.st1[-1]

    def getMin(self) -> int:
        return self.st2[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()